File: delightreading-server/src/main/java/com/delightreading/common/JsonBasedType.java
package com.delightreading.common;

import com.delightreading.SpringApplicationContextUtil;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.hibernate.HibernateException;
import org.hibernate.engine.spi.SharedSessionContractImplementor;
import org.hibernate.usertype.UserType;

import java.io.*;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;

public abstract class JsonBasedType implements UserType {

    private TypeReference deserializationTypeRef;

    public JsonBasedType(TypeReference deserializationTypeRef) {
        this.deserializationTypeRef = deserializationTypeRef;
    }
    public JsonBasedType() {
        this(null);
    }

    @Override
    public int[] sqlTypes() {
        return new int[]{Types.JAVA_OBJECT};
    }

    @Override
    public Object nullSafeGet(final ResultSet rs, final String[] names, final SharedSessionContractImplementor session,
                              final Object owner) throws HibernateException, SQLException {
        final String cellContent = rs.getString(names[0]);
        if (cellContent == null) {
            return null;
        }
        try {
            final ObjectMapper mapper = SpringApplicationContextUtil.lookUpBean(ObjectMapper.class);
            if (deserializationTypeRef !=  null) {
                return mapper.readValue(cellContent.getBytes("UTF-8"), deserializationTypeRef);
            } else {
                return mapper.readValue(cellContent.getBytes("UTF-8"), returnedClass());
            }
        } catch (final Exception ex) {
            throw new RuntimeException("Failed to convert String to " + returnedClass().getSimpleName() + ": " + ex.getMessage(), ex);
        }
    }

    @Override
    public void nullSafeSet(final PreparedStatement ps, final Object value, final int idx,
                            final SharedSessionContractImplementor session) throws HibernateException, SQLException {
        if (value == null) {
            ps.setNull(idx, Types.OTHER);
            return;
        }
        try {
            final ObjectMapper mapper = SpringApplicationContextUtil.lookUpBean(ObjectMapper.class);
            final StringWriter w = new StringWriter();
            mapper.writeValue(w, value);
            w.flush();
            ps.setObject(idx, w.toString(), Types.OTHER);
        } catch (final Exception ex) {
            throw new RuntimeException("Failed to convert " + returnedClass().getSimpleName() + "to String : " + ex.getMessage(), ex);
        }
    }

    @Override
    public Object deepCopy(final Object value) throws HibernateException {
        try {
            // use serialization to create a deep copy
            ByteArrayOutputStream bos = new ByteArrayOutputStream();
            ObjectOutputStream oos = new ObjectOutputStream(bos);
            oos.writeObject(value);
            oos.flush();
            oos.close();
            bos.close();

            ByteArrayInputStream bais = new ByteArrayInputStream(bos.toByteArray());
            return new ObjectInputStream(bais).readObject();
        } catch (ClassNotFoundException | IOException ex) {
            throw new HibernateException(ex);
        }
    }

    @Override
    public boolean isMutable() {
        return false;
    }

    @Override
    public Serializable disassemble(Object o) {
        return (Serializable) o;
    }

    @Override
    public Object assemble(Serializable cached, Object owner) {
        return cached;
    }

    @Override
    public Object replace(Object o, Object target, Object owner) {
        return o;
    }

    @Override
    public boolean equals(Object x, Object y) {
        return (x == y) || (x != null && x.equals(y));
    }

    @Override
    public int hashCode(Object x) {
        return x.hashCode();
    }
}

File: delightreading-backend/src/main/kotlin/com/delightreading/common/JsonBasedType.kt
package com.delightreading.common

import com.delightreading.SpringApplicationContextUtil
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import org.hibernate.HibernateException
import org.hibernate.engine.spi.SharedSessionContractImplementor
import org.hibernate.usertype.UserType
import java.io.*
import java.nio.charset.StandardCharsets
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.SQLException
import java.sql.Types

abstract class JsonBasedType(val deserializationTypeRef: TypeReference<*>? = null): UserType {

    override fun sqlTypes(): IntArray {
        return intArrayOf(Types.JAVA_OBJECT)
    }

    @Throws(HibernateException::class, SQLException::class)
    override fun nullSafeGet(
        rs: ResultSet, names: Array<String>, session: SharedSessionContractImplementor,
        owner: Any
    ): Any? {
        val cellContent = rs.getString(names[0]) ?: return null
        try {
            val mapper = SpringApplicationContextUtil.lookUpBean(ObjectMapper::class.java)
            return if (deserializationTypeRef != null) {
                mapper.readValue(cellContent.toByteArray(StandardCharsets.UTF_8), deserializationTypeRef)
            } else {
                mapper.readValue(cellContent.toByteArray(StandardCharsets.UTF_8), returnedClass())
            }
        } catch (ex: Exception) {
            throw RuntimeException("Failed to convert String to " + returnedClass().simpleName + ": " + ex.message, ex)
        }

    }

    @Throws(HibernateException::class, SQLException::class)
    override fun nullSafeSet(
        ps: PreparedStatement, value: Any?, idx: Int,
        session: SharedSessionContractImplementor
    ) {
        if (value == null) {
            ps.setNull(idx, Types.OTHER)
            return
        }
        try {
            val mapper = SpringApplicationContextUtil.lookUpBean(ObjectMapper::class.java)
            val w = StringWriter()
            mapper.writeValue(w, value)
            w.flush()
            ps.setObject(idx, w.toString(), Types.OTHER)
        } catch (ex: Exception) {
            throw RuntimeException("Failed to convert " + returnedClass().simpleName + " to String : " + ex.message, ex)
        }

    }

    @Throws(HibernateException::class)
    override fun deepCopy(value: Any?): Any? {
        try {
            // use serialization to create a deep copy
            val bos = ByteArrayOutputStream()
            val oos = ObjectOutputStream(bos)
            oos.writeObject(value)
            oos.flush()
            oos.close()
            bos.close()

            val bais = ByteArrayInputStream(bos.toByteArray())
            return ObjectInputStream(bais).readObject()
        } catch (ex: ClassNotFoundException) {
            throw HibernateException(ex)
        } catch (ex: IOException) {
            throw HibernateException(ex)
        }

    }

    override fun isMutable(): Boolean {
        return false
    }

    override fun disassemble(o: Any): Serializable {
        return o as Serializable
    }

    override fun assemble(cached: Serializable, owner: Any): Any {
        return cached
    }

    override fun replace(o: Any, target: Any, owner: Any): Any {
        return o
    }

    override fun equals(x: Any?, y: Any?): Boolean {
        return x === y || x != null && x == y
    }

    override fun hashCode(x: Any): Int {
        return x.hashCode()
    }
}